### ISSUE TITLE
Please fix 1.2.12 compile

### ISSUE THREAD

**Issue by Spatzendame** [madler/zlib#631]
Hi Mr. Adler,

I can't build zlibwapi.dll from your new version 1.2.12 because of errors during the compiling process using Visual Studio v17 2022. The previous version 1.2.11 worked.
I see many people have this problem. Can you please make a new release?

Thank you so much.

**Comment by topwizprogramming** [comment]
I have the same problem in VS 2012 and VS 2019. First attempt fails due to missing bat file that is listed as a pre-build event. After removing the bat file, it gives a large number of errors. I have attached the output from VS 2012.

[compile_errors.txt](https://github.com/madler/zlib/files/8455925/compile_errors.txt)

**Comment by tbeu** [comment]
@topwizprogramming See #624 and #620 for the proposed fixes.

**Comment by Spatzendame** [comment]
I want a native fix and not workaround solutions.

**Comment by ied206** [comment]
## Summary

`contrib/masmx86` and `contrib/masmx64` was removed in 1.2.12 release.
It is most likely due to #274, which was a bug that x86 assembly optimization broke the library.
However, MSVC project files under `contrib/vstudio` were not updated for this change.

## Details

### `inffas8664.c` issue

`contrib/masmx64` contained optimized version of `inffast.c`, `inffas8664.c`.
On x64 build, this source is always referred to even if you build with the `ReleaseWithoutAsm` configuration.
It is not present on 1.2.11, so every x64 build fails.

```xml
<ClCompile Include="..\..\masmx64\inffas8664.c">
  <ExcludedFromBuild Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">true</ExcludedFromBuild>
  <ExcludedFromBuild Condition="'$(Configuration)|$(Platform)'=='ReleaseWithoutAsm|Win32'">true</ExcludedFromBuild>
  <ExcludedFromBuild Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">true</ExcludedFromBuild>
</ClCompile>
```

### Assembly optimization issue

> First attempt fails due to missing bat file that is listed as a pre-build event

If built with `Release` config, zlib MSVC projects try to run masm in `contrib/masmx86` or `contrib/masmx64` directory.

```xml
<PreBuildEvent>
  <Command>cd ..\..\contrib\masmx64
bld_ml64.bat</Command>
</PreBuildEvent>
```

These directories were deleted in 1.2.12, thus the build fails.

**Comment by ied206** [comment]
## x86 code error

I also noticed on the x86 builds with MSVC 2022, MSVC creates error on these functions:

- `crc32_combine` - [C2708](https://docs.microsoft.com/en-US/cpp/error-messages/compiler-errors-2/compiler-error-c2708?view=msvc-170)
- `crc32_combine_gen` - [C2708](https://docs.microsoft.com/en-US/cpp/error-messages/compiler-errors-2/compiler-error-c2708?view=msvc-170)
- `crc32_combine_op` - [C2373](https://docs.microsoft.com/en-US/cpp/error-messages/compiler-errors-1/compiler-error-c2373?view=msvc-170)

### C2708 - `crc32_combine` & `crc32_combine_gen` 

MSVC 2022 is not letting `z_off_t` to `z_off64_t` casting on x86 builds.
Insert explicit casting to fix this error.

```c
uLong ZEXPORT crc32_combine_gen(len2)
    z_off_t len2;
{
    // return crc32_combine_gen64(len2);
    return crc32_combine_gen64((z_off64_t)len2);
}
```

### C2373 - `crc32_combine_op`

`crc32_combine_op` is declared with `ZEXPORT` on `zlib.h`, but not on `crc32.c`.

```c
// from zlib.h
ZEXTERN uLong ZEXPORT crc32_combine_op OF((uLong crc1, uLong crc2, uLong op));
// from crc32.c
uLong crc32_combine_op(crc1, crc2, op)
```

Insert `ZEXPORT` to `crc32.c` to fix this error.
stdcall and cdecl difference is no longer relevant on non-x86 platform, so this error does not appear on x64 or arm64 build.

**Comment by Spatzendame** [comment]
Thank you with the crc32.c edits I was able to compile zlibwapi.dll and zlib.dll without errors in RleaseWithoutAsm mode.

**Comment by topwizprogramming** [comment]
I applied the above changes in VS 2019 and I am getting one error on Win32 compile:
crc32.c(1089,27): error C2708: 'crc32_combine64': actual parameters length in bytes differs from previous call or reference
crc32.c(1072): message : see declaration of 'crc32_combine64'

**Comment by topwizprogramming** [comment]
I have Visual Studio 2012, 2017, 2019, and 2022 so once the code is fixed, I can upload solution/project files for them. I've removed Itanium from them.

**Comment by Spatzendame** [comment]
Here is fixed master. I have applied all the mentioned workarounds for **_Win32 ReleaseWithoutAsm_** build mode.

https://www.mediafire.com/file/jy55bqeh5kosjly/zlib-master.zip/file

**Comment by topwizprogramming** [comment]
I created solutions/projects for Visual Studio 2017, 2019, and 2022.

I removed the Itanium platform. I also changed the Runtime Library option to /MT so that the runtime will be built-in.

https://www.topwizprogramming.com/temp/vstudio.zip

**Comment by Edward19910720** [comment]
mother fuck

**Comment by Edward19910720** [comment]
> Here is fixed master. I have applied all the mentioned workarounds for **_Win32 ReleaseWithoutAsm_** build mode.
> 
> https://www.mediafire.com/file/jy55bqeh5kosjly/zlib-master.zip/file

not work

**Comment by ramueedupalli** [comment]
Can someone upload zlib1.dll for 1.2.12 version built using VS2010.

**Comment by ramueedupalli** [comment]
> 

For me zlibwapi.dll only getting generated. How to build zlib1.dll?

**Comment by Duke98666** [comment]
> ### Workaround
> 1. Remove `<ClCompile Include="..\..\masmx64\inffas8664.c">` lines from `contrib/vstudio/vc14/zlibvc.vcxproj`.
> 2. Use `ReleaseWithoutAsm` config instead of `Release`.

With Visual Studio 2012, after making the suggested workaround in `contrib/vstudio/vc11/zlibvc.vcxproj`, I had to follow this issue #668. I redeclared `(void)opaque;` at the end of all variable declarations in `contrib/minizip/ioapi.c`. With that I was able to build the zlibvc project successfully with x64 platform.

**Comment by tbeu** [comment]
Resolved by 7ecf7c7458578d05a20fa481436dd5c58db112f7 and 6c532a8e8a2fcedc4abbe2378dd26b5d89cf1c4a.

**Comment by madler** [comment]
Per @tbeu 's assertion, I will close this issue.
